libnymea-core/modbus/modbusrtumasterimpl.cpp

```diff
@@ -0,0 +1,513 @@
+        qCDebug(dcModbusRtu()) << "Connection state changed" << m_modbusUuid.toString() << m_serialPort << state;
+        if (state == QModbusDevice::ConnectedState) {
+            if (m_connected != true) {
+                m_connected = true;
+                emit connectedChanged(m_connected);
+            }
+        } else {
+            if (m_connected != false) {
+                m_connected = false;
+                emit connectedChanged(m_connected);
+            }
+        }
+    });
+
+    connect(m_modbus, &QModbusRtuSerialMaster::errorOccurred, this, [=](QModbusDevice::Error error){
+        qCWarning(dcModbusRtu()) << "Error occured for modbus RTU master" << m_modbusUuid.toString() << m_serialPort << error << m_modbus->errorString();
+        if (error != QModbusDevice::NoError) {
+            disconnectDevice();
+        }
+    });
+#endif
+}
+
+QUuid ModbusRtuMasterImpl::modbusUuid() const
+{
+    return m_modbusUuid;
+}
+
+QString ModbusRtuMasterImpl::serialPort() const
+{
+    return m_serialPort;
+}
+
+void ModbusRtuMasterImpl::setSerialPort(const QString &serialPort)
+{
+    if (m_serialPort == serialPort)
+        return;
+
+    m_serialPort = serialPort;
+    emit serialPortChanged(m_serialPort);
+}
+
+qint32 ModbusRtuMasterImpl::baudrate() const
+{
+    return m_baudrate;
+}
+
+void ModbusRtuMasterImpl::setBaudrate(qint32 baudrate)
+{
+    if (m_baudrate == baudrate)
+        return;
+
+    m_baudrate = baudrate;
+    emit baudrateChanged(m_baudrate);
+}
+
+QSerialPort::Parity ModbusRtuMasterImpl::parity() const
+{
+    return m_parity;
+}
+
+void ModbusRtuMasterImpl::setParity(QSerialPort::Parity parity)
+{
+    if (m_parity == parity)
+        return;
+
+    m_parity = parity;
+    emit parityChanged(m_parity);
+}
+
+QSerialPort::DataBits ModbusRtuMasterImpl::dataBits() const
+{
+    return m_dataBits;
+}
+
+void ModbusRtuMasterImpl::setDataBits(QSerialPort::DataBits dataBits)
+{
+    if (m_dataBits == dataBits)
+        return;
+
+    m_dataBits = dataBits;
+    emit dataBitsChanged(m_dataBits);
+}
+
+QSerialPort::StopBits ModbusRtuMasterImpl::stopBits()
+{
+    return m_stopBits;
+}
+
+void ModbusRtuMasterImpl::setStopBits(QSerialPort::StopBits stopBits)
+{
+    if (m_stopBits == stopBits)
+        return;
+
+    m_stopBits = stopBits;
+    emit stopBitsChanged(m_stopBits);
+}
+
+bool ModbusRtuMasterImpl::connected() const
+{
+    return m_connected;
+}
+
+bool ModbusRtuMasterImpl::connectDevice()
+{
+#ifdef WITH_QTSERIALBUS
+    m_modbus->setConnectionParameter(QModbusDevice::SerialPortNameParameter, m_serialPort);
+    m_modbus->setConnectionParameter(QModbusDevice::SerialBaudRateParameter, m_baudrate);
+    m_modbus->setConnectionParameter(QModbusDevice::SerialDataBitsParameter, m_dataBits);
+    m_modbus->setConnectionParameter(QModbusDevice::SerialStopBitsParameter, m_stopBits);
+    m_modbus->setConnectionParameter(QModbusDevice::SerialParityParameter, m_parity);
+    return m_modbus->connectDevice();
+#else
```

Comment:
I wonder if it wouldn't make sense to ifdef this at a higher level, i.e. not even build this stuff if it can't work anyways.... Not sure what other ramifications that might have though... if that would cause moving all those ifdefs into the json handler we wouldn't win too much... But perhaps it could make sense to not even register the Modbus json interface in that case? dunno, just a thought...